`storage`: fixes for book keeping of `dirty` and `closed` bytes

WillemKauf opened the pull request:
There were a number of bugs in the `disk_log_impl` `dirty` and `closed` bytes book-keeping added in PR https://github.com/redpanda-data/redpanda/pull/24649 that are fixed by this PR:

1. Some calls to `subtract_dirty_segment_bytes` which did not check if the segment in question was already cleanly compacted. This would leave to over-deduction of dirty bytes.
2. In `do_compact_adjacent_segments()`, it is possible to merge a cleanly compacted segment with a non-cleanly compacted segment. Later on, when this replacement segment is finally marked cleanly compacted, the sum of bytes (including those that have already been removed when the cleanly compacted segment was marked) will once again be deducted. We must add these back and consider them as "dirty" to avoid this over-deduction.
3. Also in `do_compact_adjacent_segments()`, a call to `remove_segment_permanently()` for one of the segments being merged would remove its dirty and closed bytes. These bytes are still present in the replacement segment and should not be removed.
4. Segments removed in `remove_segment_permanently()` should only deduct bytes iff the segment was previously closed (i.e does not have an appender).
5. The reduction in bytes caused by a truncation operation was not deducted from the dirty/closed bytes counter.
6. Finally, in `rewrite_segment_with_offset_map()`, we would again over-deduct dirty bytes for a segment that was already marked as cleanly compacted.

Another major issue was the assumption that, for a compaction result, `size_after < size_before`. This is an invalid assumption, as we recompress batches during the compaction process, which could lead to a larger segment post compaction.

To fix the resulting issue of possible unsigned integer underflow in this case, this PR replaces the types of book-keeping variables with their signed integer equivalents.

Finally, a fixture test that utilizes a fixed number of random operations which can influence the dirty and closed bytes of a log and compares the expected values with the book-kept values after each operation is added to ensure correctness.


## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

* none

WillemKauf committed:
`storage`: add `segment::has_clean_compact_timestamp()`

No functional changes in this commit, just a new accessor function.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--b3adc6d0-5be6-4b0d-9e01-db44675f9ae5-->
<!--TEST RESULTS-->
<!--dfd92fdd-4e1d-49a7-bba4-1a382d3a9840-->
<!--TEST RESULTS-->
<!--c611e993-f796-403b-bbb8-0386eb245f46-->
<!--TEST RESULTS-->
<!--1c324b52-c30e-4354-8254-0c48245627fc-->
<!--TEST RESULTS-->
<!--bf1d72ba-2937-4caa-8cd1-810b79ff823d-->
<!--TEST RESULTS-->
<!--2a082a43-edaf-490f-98dc-2e088cf5f77b-->
<!--TEST RESULTS-->
<!--aeba7cf9-21ac-4d1d-9826-b24112df9937-->
<!--TEST RESULTS-->
<!--45122237-7d23-4e14-902a-baa71f65049f-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61806">build#61806</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.compaction_test.CompactionGapsTest.test_translation_no_gaps.cloud_storage_type=CloudStorageType.S3.catalog_type=CatalogType.REST_HADOOP | ducktape | https://buildkite.com/redpanda/redpanda/builds/61806#0194f760-a7a1-4466-b888-e765d2454312 | FLAKY | 1/2 |
| rptest.tests.e2e_shadow_indexing_test.ShadowIndexingWhileBusyTest.test_create_or_delete_topics_while_busy.short_retention=True.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/61806#0194f760-a7a1-4466-b888-e765d2454312 | FLAKY | 1/2 |
| rptest.transactions.producers_api_test.ProducersAdminAPITest.test_producers_state_api_during_load | ducktape | https://buildkite.com/redpanda/redpanda/builds/61806#0194f773-2fb8-4b10-be0c-7c7dedf56454 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61882">build#61882</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61882#01950739-d8b0-4788-9301-c192817deb87 | FLAKY | 1/3 |
| rptest.transactions.consumer_offsets_test.VerifyConsumerOffsets.test_consumer_group_offsets | ducktape | https://buildkite.com/redpanda/redpanda/builds/61882#0195073e-1888-4030-ac43-478a8e116c57 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61901">build#61901</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| opfuzz_test_rpunit.opfuzz_test_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/61901#01951195-ced4-4c41-b47c-db587aa29fc8 | FLAKY | 1/2 |
| rptest.tests.archival_test.ArchivalTest.test_all_partitions_leadership_transfer.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61901#019511da-000b-4418-899f-9c4faed9b952 | FLAKY | 1/2 |
| rptest.tests.archival_test.ArchivalTest.test_single_partition_leadership_transfer.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61901#019511da-000b-4418-899f-9c4faed9b952 | FLAKY | 1/5 |
| rptest.tests.cluster_recovery_test.ClusterRecoveryTest.test_bootstrap_with_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61901#019511da-000b-4418-899f-9c4faed9b952 | FAIL | 0/20 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61901#019511f5-246f-4de1-8f58-48e608376510 | FLAKY | 1/2 |
| rptest.tests.delete_records_test.DeleteRecordsTest.test_delete_records_segment_deletion.cloud_storage_enabled=False.truncate_point=at_segment_boundary | ducktape | https://buildkite.com/redpanda/redpanda/builds/61901#019511da-000b-4418-899f-9c4faed9b952 | FLAKY | 1/17 |
| rptest.tests.leadership_transfer_test.MultiTopicAutomaticLeadershipBalancingTest.test_topic_aware_rebalance | ducktape | https://buildkite.com/redpanda/redpanda/builds/61901#019511da-000b-4418-899f-9c4faed9b952 | FAIL | 0/20 |
| rptest.tests.prefix_truncate_recovery_test.PrefixTruncateRecoveryTest.test_prefix_truncate_recovery.acks=1.start_empty=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/61901#019511da-000b-4418-899f-9c4faed9b952 | FAIL | 0/20 |
| rptest.tests.shard_placement_test.ShardPlacementTest.test_node_join.disable_license=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61901#019511da-000b-4418-899f-9c4faed9b952 | FLAKY | 1/2 |
| rptest.tests.upgrade_test.UpgradeWithWorkloadTest.test_rolling_upgrade_with_rollback.upgrade_after_rollback=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/61901#019511da-000b-4418-899f-9c4faed9b952 | FLAKY | 1/2 |
| storage_e2e_single_thread_rpunit.storage_e2e_single_thread_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/61901#01951195-ced4-4c41-b47c-db587aa29fc8 | FAIL | 0/2 |
| test_raft_rpunit.test_raft_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/61901#01951195-ced4-4c41-b47c-db587aa29fc8 | FAIL | 0/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61949">build#61949</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_storage_rpunit.gtest_storage_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/61949#019517c6-780b-4080-b307-b7a10062fb7d | FAIL | 0/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/61949#0195180a-cd13-4695-967a-ee3d48ae5a48 | FLAKY | 1/2 |
| rptest.tests.delete_records_test.DeleteRecordsTest.test_delete_records_concurrent_truncations.cloud_storage_enabled=True.truncate_point=random_offset | ducktape | https://buildkite.com/redpanda/redpanda/builds/61949#0195180a-cd11-47c4-a698-80ecca80bbfd | FLAKY | 1/2 |
| rptest.tests.full_disk_test.FullDiskReclaimTest.test_full_disk_triggers_gc | ducktape | https://buildkite.com/redpanda/redpanda/builds/61949#0195180a-cd12-4dfb-80f1-cd7068b6def5 | FLAKY | 1/2 |
| rptest.tests.log_segment_ms_test.SegmentMsTest.test_segment_rolling.use_alter_cfg=True.server_cfg=None.topic_cfg=300000 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61949#0195180a-cd13-4695-967a-ee3d48ae5a48 | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/61949#01951825-27dc-45cc-9481-403c0dadf09b | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61971">build#61971</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61971#01951a8b-25d9-4b38-b343-147fc2ea1527 | FLAKY | 1/4 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/61971#01951a8b-25d7-4e78-af84-d81b024fe86f | FLAKY | 1/2 |
| rptest.tests.delete_records_test.DeleteRecordsTest.test_delete_records_with_transactions.cloud_storage_enabled=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/61971#01951a71-9662-4f88-9dc7-6f6793b274df | FLAKY | 1/2 |
| rptest.tests.partition_movement_test.SIPartitionMovementTest.test_shadow_indexing.num_to_upgrade=0.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/61971#01951a8b-25d9-4b38-b343-147fc2ea1527 | FLAKY | 1/2 |
| test_archival_service_rpfixture.test_archival_service_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/61971#01951a2d-0e20-4397-8bd1-1d4710e63627 | FAIL | 0/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61980">build#61980</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/61980#01951b2e-3bce-431b-8a8c-c8d941be845d | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61980#01951b8b-f3c8-453a-b42d-63aae85453d3 | FLAKY | 1/3 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/61980#01951b8a-a14b-4f95-995e-5d03ec60aa9c | FLAKY | 1/2 |
| rptest.tests.offset_for_leader_epoch_test.OffsetForLeaderEpochTest.test_offset_for_leader_epoch | ducktape | https://buildkite.com/redpanda/redpanda/builds/61980#01951b8a-a149-4184-b4fc-f173315c01bc | FLAKY | 1/2 |
| rptest.tests.partition_force_reconfiguration_test.PartitionForceReconfigurationTest.test_basic_reconfiguration.acks=-1.restart=True.controller_snapshots=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61980#01951b8a-a14b-4f95-995e-5d03ec60aa9c | FLAKY | 1/2 |
| rptest.tests.schema_registry_test.SchemaRegistryAutoAuthTest.test_schema_id_validation.protocol=PROTOBUF.client_type=Python.validate_schema_id=True.subject_name_strategy=SubjectNameStrategyCompat.TOPIC_RECORD_NAME.payload_class=com.redpanda.CompressiblePayload.compression_type=CompressionTypes.ZSTD | ducktape | https://buildkite.com/redpanda/redpanda/builds/61980#01951b8a-a149-4184-b4fc-f173315c01bc | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61992">build#61992</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.compacted_term_rolled_recovery_test.CompactionTermRollRecoveryTest.test_compact_term_rolled_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61992#01951c5e-51e6-4a15-8b53-6e62f9e6d2a3 | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryTest.test_index_recovery | ducktape | https://buildkite.com/redpanda/redpanda/builds/61992#01951c71-bf89-48ee-b61c-e0c5ac0ba16b | FLAKY | 1/2 |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade | ducktape | https://buildkite.com/redpanda/redpanda/builds/61992#01951c71-bf8b-486c-9e83-126c628bbf2b | FLAKY | 1/2 |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape | https://buildkite.com/redpanda/redpanda/builds/61992#01951c5e-51e7-46ba-978b-6e0f53ac4141 | FLAKY | 1/3 |
| rptest.tests.timequery_test.TimeQueryTest.test_timequery_with_trim_prefix.cloud_storage=True.spillover=False | ducktape | https://buildkite.com/redpanda/redpanda/builds/61992#01951c71-bf84-4215-b418-e1fe2154307b | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62044">build#62044</a></summary>


| test_id                                                                                                                                                       | test_kind   | job_url                                                                                   | test_status   | passed   |
|:--------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.delete_records_test.DeleteRecordsTest.test_delete_records_with_transactions.cloud_storage_enabled=True                                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62044#01952144-246f-42bd-8e0d-6fb1a4b45ac7 | FLAKY         | 1/2      |
| rptest.tests.partition_force_reconfiguration_test.PartitionForceReconfigurationTest.test_basic_reconfiguration.acks=-1.restart=True.controller_snapshots=True | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62044#01952144-246f-42bd-8e0d-6fb1a4b45ac7 | FLAKY         | 1/2      |
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic                                                                               | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62044#0195215d-ba6e-4a9b-87d6-bd3ef3505494 | FLAKY         | 1/3      |
</details>
<!--/body-->

WillemKauf commented:
/ci-repeat 1
skip-build
skip-units
dt-repeat=50
tests/rptest/tests/log_compaction_test.py

nvartolomei commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1003,8 +1018,15 @@ ss::future<compaction_result> disk_log_impl::do_compact_adjacent_segments(
+
+    add_dirty_segment_bytes(dirty_bytes_to_readd);
+    const auto removed_bytes = ret.size_before - ret.size_after;
+    if (!target->has_clean_compact_timestamp()) {
```
Disclaimer: I'm still confused about dirty/clean segments but trying to make sense of this as a learning exercise too.

I'm confused by this if statement. To me it looks like a copy paste mistake. `target` here is the newly created segment. `has_clean_compact_timestamp` is the property on that new segment?

Why we're subtracting removed_bytes based on the "clean state" of the new segment? What does this mean?


> ~let's say we started with 2 segments:~
>   ~- a) clean 0 - dirty bytes; b) non-clean - 100 dirty bytes~
> 
> ~the dirty bytes counter is at 100~
> 
> ~`dirty_bytes_to_readd` is at 100 too~
> 
> ~the newly compacted segment is 75 bytes for example, therefore, removed_bytes = 25~
> 
> ~target in this case is considered non-clean? if so~
> ~we add 100 to dirty bytes (counter is at 200)~
> ~we subtract 25 * 2~
> ~counter for dirty is at 150?~
> 
> ~Questions:~
> ~* is the math right?~
> ~* are you sure you want to check target here rather than `segments.back()`?~

On the second look the math looks all right.

Last question: non-trivial to add a test for these "metrics"? The logic doesn't seem very simple so maybe we can add a test to make sure we don't break it in the future?

WillemKauf replied:
>I'm still confused about dirty/clean segments but trying to make sense of this as a learning exercise too.

A "clean" segment is a closed segment that has had all of its keys fully deduplicated in its prefix of the log. Clean segments can only be produced as a result of being fully indexed (and using the key-offset map to eliminate all duplicates in earlier segments in the log) during the sliding window process. 

A "dirty' segment is a closed segment that has not yet fulfilled the above requirement.

>Why we're subtracting removed_bytes based on the "clean state" of the new segment? What does this mean?

The new segment, `target`, is only designated as "clean" if both of the segments merged together were designated as "clean". As an aside, if segment one had `n_1` bytes and segment two had `n_2` bytes, `target` will have exactly `n_t = n_1 + n_2` bytes.

In the case that `target` is not clean (i.e dirty), after the self-compaction process of `target`, we must subtract the removed bytes from dirty segment bytes. Otherwise, we would be miscounting bytes.

Take the following example in which we have only two closed segments in the log.
Segment 1 (150 bytes, clean) and Segment 2 (100 bytes, dirty). Therefore, `_dirty_segment_bytes = 100`, `_closed_segment_bytes = 250`.

1. `S1` and `S2` are selected for adjacent compaction
2. `dirty_bytes_to_readd` is therefore 150.
3. `target` is created by merging `S1` and `S2`, size is 250 bytes, segment is designated as "dirty" due to `S2` being dirty.
4. Self compaction of `target` removes 50 bytes, size of `target` is now 200 bytes
5. We add back `dirty_bytes_to_readd`; `_dirty_segment_bytes` is now 250.
6. We subtract bytes removed, `_dirty_segment_bytes` is now `200`, `_closed_segment_bytes` is now 200.

This makes sense, as we only have one closed segment in the log, and it is dirty. When it is eventually marked as clean by sliding window compaction the entire 200 bytes will be subtracted from `_dirty_segment_bytes`, leaving it as 0, while `_closed_segment_bytes` remains 200.

I hope this explains it a bit better, please let me know if there's anything else unclear.

> Last question: non-trivial to add a test for these "metrics"? The logic doesn't seem very simple so maybe we can add a test to make sure we don't break it in the future?

There is currently checks in `log_compaction_test.py`, though not very in depth. There are also fixture tests in this PR.

The most ideal test would be a fixture test in which producing, compaction and other processes are happening, in which we could loop over the segments in a log and manually compute `dirty_segment_bytes` and `closed_segment_bytes` and then compare against the book-kept values.

I'd be happy to look at adding it to make sure the book-keeping is now bug free.

nvartolomei replied:
I was thinking something very simple, more like a unit test: create few segments of precise size, compact, check the metrics equal to expected size. I haven’t worked on this part of the codebase so don’t know what’s available in terms of testing. I’ll defer to you the decision of how and when.

nvartolomei replied:
And thanks for explanation! 🙏

nvartolomei replied:
Re the if statement: it is basically the same as saying `!all_segments_cleanly_compacted` or `bytes_to_readd > 0` but in an indirect way? If so, can we remove the if statement completely? If target is cleanly compacted then bytes to readd will be 0 and calls to subtract will have no effect.

Less cognitive load for the reader.

WillemKauf replied:
Discussed in person, but no, the if statement is required.

nvartolomei commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -2794,15 +2818,17 @@ disk_log_impl::timequery(timequery_config cfg) {
-    auto removed_segment_bytes = s->size_bytes();
-    if (!s->index().has_clean_compact_timestamp()) {
-        subtract_dirty_segment_bytes(removed_segment_bytes);
+    if (deduct_bytes) {
```
I believe this would be cleaner if we didn't have the bool flag. Instead, in the caller when we create the new segment, we can increment dirty bytes accounting for what is going to be subtracted. You already added a special case there so can be extended to cover this case as well.

We're replace [a, b] with [c] and removing [b]. [c] is all dirty. If b was dirty then increment dirty bytes by [c] size, else by [c] - [a]. Wdyt?

WillemKauf replied:
Sure, I can change this to over-add to dirty bytes, expecting the subtraction later.

WillemKauf replied:
After having written this, I'm not actually sure it makes things cleaner code-wise or thinking wise- I'm leaning towards keeping the boolean.

nvartolomei commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1003,8 +1018,15 @@ ss::future<compaction_result> disk_log_impl::do_compact_adjacent_segments(
+    }
+
+    add_dirty_segment_bytes(dirty_bytes_to_readd);
+    const auto removed_bytes = ret.size_before - ret.size_after;
```
I think `removed_bytes` can be negative 🤔 `subtract_dirty_segment_bytes` accepts an uint.

We do recompress batches during compaction. If they were produced to us with the highest compression ration but we recompress with a lower compression ratio then, I imagine, the size could become larger.

WillemKauf replied:
:thinking: I was aware we recompress, but I didn't think this would ever lead to a situation where the recompressed batch could be larger than the original compressed batch...

nvartolomei replied:
Similar behaviour can be caused by compression library upgrade. There is no guarantee that compression output is the same across different versions of compressor. Or just settings changes of course like block sizes, etc.

WillemKauf replied:
Fixed this by replacing `uint64_t` types everywhere with `ssize_t`.

WillemKauf commented:
Push to:
* Replace book-keeping types with `ssize_t` to avoid possibility of integer underflow via a compaction result where `size_after > size_before` due to compression.

WillemKauf committed:
`storage`: add `dirty_and_closed_bytes_bookkeeping` test

This test uses a fixed number of operations that influence the number of
dirty and closed bytes in the log, chosen at random.

The book-kept `_closed_segment_bytes` and `_dirty_segment_bytes` are checked
after every operation in order to ensure we are accurately book keeping
these values in `disk_log_impl`.

WillemKauf commented:
Push to:
* Add `dirty_and_closed_bytes_bookkeeping` test. This is a test that uses a fixed number of random operations which can influence the `dirty` and `closed` bytes of a log, and compares the expected values with the book-kept values after each operation.

nvartolomei commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -4099,23 +4128,23 @@ double disk_log_impl::dirty_ratio() const {
-void disk_log_impl::subtract_dirty_segment_bytes(uint64_t bytes) {
-    _dirty_segment_bytes -= std::min(bytes, _dirty_segment_bytes);
+void disk_log_impl::subtract_dirty_segment_bytes(ssize_t bytes) {
+    _dirty_segment_bytes = std::max(0L, _dirty_segment_bytes - bytes);
```
This is more of a fail-safe? Do we know a scenario where this could go negative without the `max(0, ...)`? If no, I'd suggest adding a debug assertion so we can catch it in CI if it ever happens.

WillemKauf replied:
I was trying to reason about this, there shouldn't be a good reason bar a bug for this to go negative, so it is a failsafe.

Converting this to a `dassert` is a good idea I think.

WillemKauf commented:
Force push to:
* Use `dassert()` in `disk_log_impl::subtract_dirty/closed_segment_bytes()` to catch any negative values in CI.

WillemKauf commented:
Force push to:
* Rebase to `upstream/dev`

WillemKauf commented:
Argh. here be bugs. investigating

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/61901-->
<!--cc1161d0-22cd-4fd9-aac4-dd8f08ef1847-->
<!--title-->
## Retry command for [Build#61901](https://buildkite.com/redpanda/redpanda/builds/61901)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cluster_recovery_test.py::ClusterRecoveryTest.test_bootstrap_with_recovery
tests/rptest/tests/prefix_truncate_recovery_test.py::PrefixTruncateRecoveryTest.test_prefix_truncate_recovery@{"acks":1,"start_empty":true}
tests/rptest/tests/leadership_transfer_test.py::MultiTopicAutomaticLeadershipBalancingTest.test_topic_aware_rebalance
```
<!--/body-->

WillemKauf committed:
`storage`: fix typo in `storage_e2e_test`

`compation` -> `compaction`.

WillemKauf committed:
`storage`: add book-keeping checks to `storage_test_fixture`

For ensuring `dirty/closed_segment_bytes()` are correct after
various log operations.

WillemKauf commented:
Force push to:
* Hopefully fix everything

WillemKauf committed:
`storage`: deal with dirty/closed bytes in `segment_deduplication_test`

This function manually releases an appender (i.e rolls the segment produced),
so it must also correctly book-keep the dirty/closed bytes here.

WillemKauf committed:
`storage`: make fixes to `dirty_bytes` book-keeping

There were a _quite_ a number of bugs in the book-keeping that are fixed by
this commit:

1. Some calls to `subtract_dirty_segment_bytes` which did not check
   if the segment in question was already cleanly compacted. This
   would leave to over-deduction of dirty bytes.
2. In `do_compact_adjacent_segments()`, it is possible to merge
   a cleanly compacted segment with a non-cleanly compacted segment.
   Later on, when this replacement segment is finally marked cleanly
   compacted, the sum of bytes (including those that have already been
   removed when the cleanly compacted segment was marked) will once
   again be deducted. We must add these back and consider them as
   "dirty" to avoid this over-deduction.
3. Also in `do_compact_adjacent_segments()`, a call to
   `remove_segment_permanently()` for one of the segments being merged
   would remove its dirty and closed bytes. These bytes are still
   present in the replacement segment and should not be removed.
4. Segments removed in `remove_segment_permanently()` should only deduct bytes
   iff the segment was previously closed (i.e does not have an appender).
5. The reduction in bytes caused by a truncation operation was not deducted
   from the dirty/closed bytes counter.
6. Finally, in `rewrite_segment_with_offset_map()`, we would again
   over-deduct dirty bytes for a segment that was already marked as
   cleanly compacted.

WillemKauf committed:
`storage`: add `disk_log_impl::add/subtract_segment_bytes()`

WillemKauf committed:
`storage`: make `dirty/closed_segment_bytes()` virtual in `log`

For use in tests.

WillemKauf committed:
`storage`: replace book-keeping types with `ssize_t`

The assumption that the result from a compaction operation will always
result in a positive number (i.e `bytes_after < bytes_before`) was an
invalid one.

For sake of ease and correctness, replace the book-kept variables
(previously `uint64_t`) with their signed equivalents (`ssize_t`).

Also, add a `dassert()` to `subtract_closed/dirty_bytes()` to ensure
their values do not become negative.

WillemKauf commented:
Force push:
* Clean up commit history

WillemKauf commented:
Force push to:
* Reduce number of operations in `dirty_and_closed_bytes_bookkeeping` test and bump `storage_e2e_single_thread` timeout in bazel `BUILD` file to avoid timeouts in CI.

andrwng commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -519,10 +513,9 @@ ss::future<> disk_log_impl::adjacent_merge_compact(
-            subtract_dirty_segment_bytes(removed_bytes);
-            subtract_closed_segment_bytes(removed_bytes);
-
+            const ssize_t removed_bytes = ssize_t(result.size_before)
+                                          - ssize_t(result.size_after);
```
Maybe the primitive should be `replace_segment_bytes(segment, uint64_t size_before, uint64_t size_after)`? Then I'm not sure we'd even need add/subtract primitives in disk_log_impl (adding/subtracting is then just a matter of supplying 0 for one of the args), and we wouldn't need to worry about signedness at this layer

andrwng commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -875,6 +862,22 @@ ss::future<compaction_result> disk_log_impl::do_compact_adjacent_segments(
+    // compacted segment, we need to consider the bytes of the cleanly compacted
+    // segment "dirty" again, since they will be deducted once the replacement
+    // segment has finally been marked clean.
+    ssize_t dirty_bytes_to_readd = 0;
```
nit: typo

WillemKauf replied:
The intended variable name here is `readd`.

andrwng replied:
👍 consider `re_add` or something -- this seems pretty easy to mistaken for a typo

WillemKauf replied:
Changed

andrwng commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -149,13 +149,7 @@ disk_log_impl::disk_log_impl(
         if (_compaction_enabled) {
```
nit: one commit that fixes six bugs is much more difficult to review than six commits that each fix one :)

andrwng replied:
Also if the cases are mostly distinct, having a test case for each would be great too, to better understand what's being fixed

WillemKauf replied:
>nit: one commit that fixes six bugs is much more difficult to review than six commits that each fix one :)

I can split this into six commits if you'd prefer, I wasn't sure at the time if that was worth doing :) 

> Also if the cases are mostly distinct, having a test case for each would be great too, to better understand what's being fixed

I deferred to writing a randomized bookkeeping test that hits these different cases via operations that affect the dirty and closed bytes (truncation, adjacent segment merging, windowed compaction, etc) in place of attempting to write a test case for each individually.

I think it has done a very good job of catching/hitting all the bugs mentioned in this PR.

andrwng replied:
It's great to catch bugs with randomized tests. IMO we should strive to add explicit, deterministic tests when we find bugs with randomized testing

andrwng replied:
>I can split this into six commits if you'd prefer, I wasn't sure at the time if that was worth doing :)

Yea, it's almost always worth trying to optimize for reviewability/readability, especially when there are many things being touched. At this point I've spent a good amount of time looking through the PR, and I think i understand the individual lines being touched now, so I'm okay with it personally. But that's time taken from every further reviewer (and future reader) to orient themselves.

andrwng commented on src/v/storage/disk_log_impl.h:
```diff
@@ -335,7 +335,8 @@ class disk_log_impl final : public log {
 
     ss::future<> remove_segment_permanently(
       ss::lw_shared_ptr<segment> segment_to_tombsone,
-      std::string_view logging_context_msg);
+      std::string_view logging_context_msg,
+      bool deduct_bytes = true);
```
nit: a comment about the new argument would be great. It's otherwise not very clear why someone wouldn't want to deduct bytes

WillemKauf replied:
Yeah, this is currently only `false` in the call from `do_compact_adjacent_segments()`. Nicolae preferred removing the flag and instead adding the bytes to be removed pre-emptively for the same behavior.

I think I still prefer the boolean here, but I can understand that it may only be useful for this one case. I will add a comment for it.

andrwng replied:
Can I ask why you prefer adding a boolean here when it can be handled externally by the one caller using it? It doesn't feel intuitive as prospective a caller of `remove_segment_permanently`, so I feel like I'm missing something

WillemKauf replied:
Changed

andrwng commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -875,6 +862,22 @@ ss::future<compaction_result> disk_log_impl::do_compact_adjacent_segments(
+    // segment "dirty" again, since they will be deducted once the replacement
+    // segment has finally been marked clean.
+    ssize_t dirty_bytes_to_readd = 0;
+    if (!all_segments_cleanly_compacted) {
+        for (const auto& seg : segments) {
```
nit: presumably ranges also iterates through the segments. Why not just iterate through them once here in this for loop without the if block?

WillemKauf replied:
Sure, this could be written as

```cpp
bool all_segments_cleanly_compacted = true;
ssize_t dirty_bytes_to_re_add = 0;
for(const auto& seg : segments) {
    if (!seg->has_clean_compact_timestamp()) {
       all_segments_cleanly_compacted = false;
   } else {
       dirty_bytes_to_re_add += seg->size_bytes();
   }
}
```

instead if you'd prefer.

andrwng replied:
I guess i'm missing where all_segments_cleanly_compacted is used.. It doesn't appear to be used except for this if block, or am i missing it?

WillemKauf replied:
In the above change, we'd also have to amend the later part of the code

```cpp
    add_dirty_segment_bytes(dirty_bytes_to_re_add);
```

with 

```cpp
   if(!all_segments_cleanly_compacted) {
      add_dirty_segment_bytes(dirty_bytes_to_re_add);
   }
```

Currently `all_segments_cleanly_compacted` is only used here to guarantee that `dirty_bytes_to_re_add == 0` iff `all_segments_cleanly_compacted`.

https://github.com/redpanda-data/redpanda/blob/f86c1dd0f48490ceda1553baa450cc35aee80e23/src/v/storage/disk_log_impl.cc#L1012

andrwng replied:
Ahh got it. Makes sense, thanks for explaining

andrwng commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -875,6 +862,22 @@ ss::future<compaction_result> disk_log_impl::do_compact_adjacent_segments(
+    const bool all_segments_cleanly_compacted = std::ranges::all_of(
+      segments, &segment::has_clean_compact_timestamp);
+
+    // In the case we are merging a cleanly compacted segment with a non-cleanly
+    // compacted segment, we need to consider the bytes of the cleanly compacted
+    // segment "dirty" again, since they will be deducted once the replacement
+    // segment has finally been marked clean.
```
I'm having a hard time following this and it feels kind of off. Could you give an example?

WillemKauf replied:
Say we are performing adjacent segment merging with [A] and [B], where [A] is cleanly compacted (therefore has had its bytes deducted from `dirty_segment_bytes` previously) and [B] is not.


After merging [A] and [B] to target [C] (which is of size `[A] + [B]`), target segment [C] will not be marked as "cleanly compacted" due to [B], which means later on, when target [C] IS marked cleanly compacted and has all its bytes removed from `dirty_segment_bytes` during the sliding window process, we will have overly-deducted `[A]` (total deducted: `2*[A]`) dirty bytes (assuming no further compaction).

Therefore, we need to add the bytes back from cleanly compacted segment [A] when producing a target [C] which won't be marked as cleanly compacted.

andrwng replied:
I'm thinking actually maybe this is worth calling `clean_turning_dirty_bytes` or somesuch. Then it's a bit clearer conceptually what it is, rather than being described mechanically, and it'd be IMO easier to grok why we're accounting in the way we are.

andrwng reviewed (commented):
I haven't fully re-cached all the context around bookkeeping in disk_log_impl. Given how many "bugs" this PR is addressing, some explicit testing of the fixed bugs would be very helpful as a reviewer

WillemKauf committed:
`archival`: fix `dirty/closed_bytes` bookkeeping after `release_appender()`

Without manually incrementing the `dirty/closed_bytes` after forcing an appender
release, we may trigger the `dassert()` for `dirty/closed_bytes < 0`.

WillemKauf commented:
Force push to:
* Fix the "last" ( :crossed_fingers: ) of the broken fixture/unit tests which manually release segment appenders/roll segments without incrementing `closed/dirty_segment_bytes`, leading to `dassert()` failures.
* Change variable name `dirty_bytes_to_readd` -> `dirty_bytes_to_re_add`
* Remove boolean `deduct_bytes` flag from `remove_segment_permanently()`.

andrwng commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -1002,10 +1005,21 @@ ss::future<compaction_result> disk_log_impl::do_compact_adjacent_segments(
+        // remove_segment_permanently() will remove bytes, so to make sure the
+        // book-keeping stays correct, we must first add them here to ensure a
+        // net zero change post adjacent merge.
+        add_segment_bytes(segments.back(), segments.back()->size_bytes());
```
This feels off because we seem to actually be removing segment, and not adding segments in the code below. I think this would be more intuitive if we did this right after transfer_segment, since that code is actually adding bytes temporarily, to be removed here.

WillemKauf replied:
I put this here so it would be within the same `if` statement as the removal of the segment.

https://github.com/redpanda-data/redpanda/blob/9e7b251824e06d2bd0870827ad1e3b5181756aa5/src/v/storage/disk_log_impl.cc#L995-L1006

It seems correct to place the addition underneath the same safe guarding against the mentioned race condition highlighted in the linked comment, otherwise we could be over-adding to our counters if we hit the race.

andrwng replied:
Wouldn't such a prefix truncation fiber reduce the size when it calls `remove_segment_permanently()`, such that we don't over-add? Or is this comment implying that there's a case where we might remove the prefix truncation might remove the segment but forget to decrease the bytes?

andrwng replied:
If the latter, it'd be nice to address that race in a different way if we can. IMO this code placement feels off if you don't know about that race.

WillemKauf replied:
> Wouldn't such a prefix truncation fiber reduce the size when it calls remove_segment_permanently(), such that we don't over-add? 

Hm yeah. I was thinking about this wrong.

It does make more sense to move the call to `add_segment_bytes()` below `transfer_segment()`.

Going to push changes after letting CI run its course.

WillemKauf commented:
CI is green! Force pushing to:
* rename `dirty_bytes_to_re_add` to `clean_turning_dirty_bytes`
* move `add_segment_bytes()` call below `transfer_segment()` call site in `do_compact_adjacent_segments()`

WillemKauf commented:
I am on the fence about committing the `dassert()` into `dev`- on one hand, it's great for catching edge cases and bugs, but on the other, people writing tests may run into issues at the storage layer if they aren't careful about manual book-keeping, and it is also a very easy CI blocker if there (god forbid) should still be races or bugs with book-keeping. Looking for opinions here.

WillemKauf commented:
/ci-repeat 3
skip-build
skip-units
dt-repeat=10

nvartolomei commented:
> I am on the fence about committing the dassert() into dev- on one hand, it's great for catching edge cases and bugs, but on the other, people writing tests may run into issues at the storage layer if they aren't careful about manual book-keeping, and it is also a very easy CI blocker if there (god forbid) should still be races or bugs with book-keeping. Looking for opinions here.

I'm all in for keeping/working towards having the `dassert`. What you highlight here is a design problem with disk log/disk log builder. Higher level users shouldn't be aware/have exposure to the internal bookkeeping needed. Take offsets as an analogy. No callers have to advance/rewind offsets during any operations. The same should be true for dirty/closed bytes.

WillemKauf commented:
/ci-repeat 1
skip-build
skip-units

WillemKauf commented:
The failed ducktape jobs from previous `ci-repeat` aren't related to PR changes, seems that the agents were killed (infra issue)

WillemKauf commented:
CI green.

dotnwat commented on src/v/storage/disk_log_impl.cc:
```diff
@@ -4099,23 +4099,28 @@ double disk_log_impl::dirty_ratio() const {
-    _closed_segment_bytes -= std::min(bytes, _closed_segment_bytes);
+void disk_log_impl::subtract_closed_segment_bytes(ssize_t bytes) {
+    _closed_segment_bytes -= bytes;
+    dassert(
+      _closed_segment_bytes >= 0,
+      "_closed_segment_bytes must not be negative.");
```
should we log some kind of notice about this situation and/or "disable" functionality that depends on the now-invalid statistics?

WillemKauf replied:
I think another safety hatch would help here.

FWIW, I am happy with the test coverage we have at this point to ensure we're not hitting invalid negative states, and I'm happy to continue to add tests and stress it more.

We probably want the behaviour that any bogus negative values here would make the dirty ratio over-prescriptive rather than under-prescriptive when used for scheduling compaction?

In `disk_log_impl::dirty_ratio()`, we could choose to return `1` (to be over-prescriptive) instead of any negative value.

WillemKauf replied:
Actually, an even better idea- why don't we "reset" the byte counter by simply performing the `O(n)` summation operation over the segments in the log?

WillemKauf replied:
Something like:


```diff
diff --git a/src/v/storage/disk_log_impl.cc b/src/v/storage/disk_log_impl.cc
index fdb9694c08..47cea3ff36 100644
--- a/src/v/storage/disk_log_impl.cc
+++ b/src/v/storage/disk_log_impl.cc
@@ -4117,6 +4117,19 @@ ss::future<> disk_log_impl::remove_kvstore_state(
 }
 
 double disk_log_impl::dirty_ratio() const {
+    // Sanity check against negative values.
+    if (_dirty_segment_bytes < 0 || _closed_segment_bytes < 0) {
+        vlog(stlog.debug,
+          "{}: _dirty_segment_bytes ({}) and/or _closed_segment_bytes ({}) "
+          "have become negative - clearing counters and resumming over {} "
+          "segments in the log.",
+          config().ntp(),
+          _dirty_segment_bytes,
+          _closed_segment_bytes,
+          _segs.size());
+
+        reset_dirty_and_closed_bytes();
+    }
     // Avoid division by 0.
     return (_closed_segment_bytes == 0)
              ? 0.0
@@ -4165,4 +4178,19 @@ void disk_log_impl::subtract_segment_bytes(
     subtract_closed_segment_bytes(bytes);
 }
 
+void disk_log_impl::reset_dirty_and_closed_bytes() {
+    ssize_t dirty{0};
+    ssize_t closed{0};
+    for (const auto& seg : _segs) {
+        if (!seg->has_appender()) {
+            if (!seg->has_clean_compact_timestamp()) {
+                dirty += seg->file_size();
+            }
+            closed += seg->file_size();
+        }
+    }
+    _dirty_segment_bytes = dirty;
+    _closed_segment_bytes = closed;
+}
+
 } // namespace storage
diff --git a/src/v/storage/disk_log_impl.h b/src/v/storage/disk_log_impl.h
index 1393eb8bc8..8ea9e63f33 100644
--- a/src/v/storage/disk_log_impl.h
+++ b/src/v/storage/disk_log_impl.h
@@ -504,6 +504,12 @@ private:
     // closed_segment_bytes, and conditionally removed from dirty_segment_bytes.
     void subtract_segment_bytes(ss::lw_shared_ptr<segment> s, ssize_t bytes);
 
+    // Performs a manual O(n) reset of the dirty and closed bytes in the log by
+    // iterating over segment in the log. Used as a safety hatch in
+    // dirty_ratio() to prevent returning a bogus value due to improper
+    // book-keeping.
+    void reset_dirty_and_closed_bytes();
+
     bool _compaction_enabled;
 };
```

We could even go as far as to add a probe metric for # of times `reset_dirty_and_closed_bytes()` was called. But this seems like a lot of effort for something that should probably just be bug free :smile:

WillemKauf replied:
FWIW, I don't think this approach is awful, it seems like kafka re-computes this stuff on the fly as well: https://github.com/apache/kafka/blob/3fc103b48b65fc37d8ae22d4d7c36104ae1b3b35/core/src/main/scala/kafka/log/LogCleanerManager.scala#L678-L684

dotnwat replied:
Yeh i think recomputing on the fly is the easiest to make robust provided it isn't too costly.

> We probably want the behaviour that any bogus negative values here would make the dirty ratio over-prescriptive rather than under-prescriptive when used for scheduling compaction?

or the reset thing also works. i think the important probably is the (1) how we react (e.g. reset) and (2) logging some info so we can _fix_ any accounting issue that pops up.

WillemKauf replied:
https://github.com/redpanda-data/redpanda/pull/25153

dotnwat reviewed (commented):
lgtm. i know we are using `dassert`, which I think is the right thing to do. however, how should we react in production if the statistics become bogus (e.g. negative)? should we log messages or even stop processing / adjust the strategy for scheduling which uses those statistics to drive behavior?

cc @andrwng

WillemKauf commented:
> should we log messages or even stop processing / adjust the strategy for scheduling which uses those statistics to drive behavior?

Responded in comment thread
